File: leetcode_cpp/209_长度最小的子数组.cpp
//
// Created by 16517 on 2024/7/10.
//
#include <vector>
#include <climits>
using namespace std;

class Solution {
public:
    int minSubArrayLen(int s, vector<int>& nums) {
        int ans = INT_MAX, sum = 0;
        int left = 0, right = 0;
        while (right < nums.size()) {
            if (sum < s) {
                sum += nums[right++];
            }
            while (sum > s) {
                sum -= nums[left++];
            }
            if (sum == s) {
                ans = ans < (right - left + 1) ? ans : (right - left + 1);
                right++;
            }
        }
        return ans;
    }
};

File: leetcode_rust/src/209_长度最小的子数组.rs
struct Solution {

}

impl Solution {
    pub fn min_sub_array_len(target: i32, nums: Vec<i32>) -> i32 {
        let mut ans: usize = 0;
        let mut sum: i32 = 0;
        let mut left:usize = 0;
        for right in 0..nums.len() {
            sum += nums[right];
            while sum >= target {
                if ans == 0 || ans > (right - left + 1) {
                    ans = right - left + 1;
                }
                sum -= nums[left];
                left += 1;
            }
        }
        return ans as i32
    }
}